iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/unique_ptr.inl

```diff
@@ -45,67 +45,67 @@ inline unique_ptr<T>& unique_ptr<T>::operator=(unique_ptr&& rhs) noexcept
     return *this;
 }
 
-template <typename T>
-inline unique_ptr<T>::unique_ptr(unique_ptr&& rhs) noexcept
+template <typename T, typename D>
+inline unique_ptr<T, D>::unique_ptr(unique_ptr&& rhs) noexcept
     : m_ptr{rhs.m_ptr}
     , m_deleter{std::move(rhs.m_deleter)}
 {
     rhs.m_ptr = nullptr;
 }
 
-template <typename T>
-inline unique_ptr<T>::~unique_ptr() noexcept
+template <typename T, typename D>
+inline unique_ptr<T, D>::~unique_ptr() noexcept
 {
     destroy();
 }
 
-template <typename T>
-inline T* unique_ptr<T>::operator->() noexcept
+template <typename T, typename D>
+inline T* unique_ptr<T, D>::operator->() noexcept
 {
     cxx::Expects(m_ptr != nullptr);
     return get();
 }
 
-template <typename T>
-inline const T* unique_ptr<T>::operator->() const noexcept
+template <typename T, typename D>
+inline const T* unique_ptr<T, D>::operator->() const noexcept
 {
     // Avoid code duplication
     // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
-    return const_cast<unique_ptr<T>*>(this)->operator->();
+    return const_cast<unique_ptr<T, D>*>(this)->operator->();
 }
 
-template <typename T>
-inline T* unique_ptr<T>::get() noexcept
+template <typename T, typename D>
+inline T* unique_ptr<T, D>::get() noexcept
 {
     return m_ptr;
 }
 
-template <typename T>
-inline const T* unique_ptr<T>::get() const noexcept
+template <typename T, typename D>
+inline const T* unique_ptr<T, D>::get() const noexcept
 {
     // AXIVION Next Construct AutosarC++19_03-A5.2.3 : Avoid code duplication
     // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
-    return const_cast<unique_ptr<T>*>(this)->get();
+    return const_cast<unique_ptr<T, D>*>(this)->get();
 }
 
-template <typename T>
-inline T* unique_ptr<T>::release(unique_ptr&& ptrToBeReleased) noexcept
+template <typename T, typename D>
+inline T* unique_ptr<T, D>::release(unique_ptr&& ptrToBeReleased) noexcept
 {
     auto ptr = ptrToBeReleased.m_ptr;
     ptrToBeReleased.m_ptr = nullptr;
     return ptr;
 }
 
-template <typename T>
-inline void unique_ptr<T>::reset(T* const ptr) noexcept
+template <typename T, typename D>
+inline void unique_ptr<T, D>::reset(T* const ptr) noexcept
```

Comment:
@elBoberido Regarding [your comment](https://github.com/eclipse-iceoryx/iceoryx/issues/1655#issuecomment-1254918642): I don't think avoiding such a misuse makes sense. It would be over-the-top. The problem in your example are the `static_cast`s and not the API of `memory::unique_ptr`. The user is responsible for providing the correct types. Furthermore, users should not use the untyped API directly, that's something to be used inside of frameworks and not directly.

Reply:
@mossmaurice it's not the `static_cast` since both of the types are the same. The problem is that the pointer must be released with a different object which is captured by the lambda. I used the casts since the `unique_ptr` cannot handle `void*` and I didn't want to create a different example. We have this situation in a lower layer though.